### PR TITLE
HOCS-4373: rework user permissions

### PR DESCRIPTION
**Pull request by LiamMacP**
Presently, a lot of our checks revolve around getting a users teams and
then calculating the case types that user is a case admin for. While
this works, if we ever need them teams we have to search for all the
teams for them case types separately. This change offers an 'expanded'
view of a users teams whereby all the team uuids that a user is a member
of, or a case admin of are added together. This will support future
permission development for ensuring users have the correct access to
workstacks and teams.

A side effect of this change is we gain a performance bonus by removing
the case type check this simplifies the queries that used it. From
`12329.683ms` to `731.688ms` on a production dataset.

### CODE REVIEW

File: src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java [hunk 1]
```diff
@@ -32,31 +33,54 @@ public UUID getUserId() {
...
+    public Set<UUID> getExpandedUserTeams() {
+        var userTeamUuids = getUserTeams();
 
+        if (userTeamUuids.isEmpty()) {
+            return userTeamUuids;
+        }
 
+        var allTeams = infoClient.getTeams();
+
+        var userCaseAdminCaseTypes = getUserCaseAdminCaseTypes(allTeams, getUserTeams());
```
Comment (nickmcmahon01) [inline src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java:65]: not blocking, but can we call getUserTeams once in this method?

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/DashboardService.java [hunk 2]
```diff
@@ -37,20 +37,13 @@ public DashboardService(DashboardSummaryFactory dashboardSummaryFactory,
     Map<UUID, Map<String, Integer>> getDashboard() {
         log.debug("Getting dashboard summary for user");
 
-        List<Summary> casesSummaries = new ArrayList<>();
-
-        Set<UUID> allUserTeams = userPermissionsService.getUserTeams();
-        if(!allUserTeams.isEmpty()) {
-            casesSummaries.addAll(summaryRepository.findTeamsAndCaseCountByTeamUuid(allUserTeams));
-        } else {
+        Set<UUID> allUserTeams = userPermissionsService.getExpandedUserTeams();
```
Comment (nickmcmahon01) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/DashboardService.java:40]: This is much simpler!

File: src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java [hunk 3]
```diff
@@ -1,25 +1,26 @@
...
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 @Service
-@Slf4j
 public class UserPermissionsService {
 
-    private RequestData requestData;
-    private InfoClient infoClient;
+    private final RequestData requestData;
```
Comment (nickmcmahon01) [inline src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java:21]: 👮 I'm glad the final police have swung by. this is the most important change in this PR IMO

File: src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java (hunk 1 above)
Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java:65]: I put it above for reuse, not sure why I didn’t change this. Will sort it later!

**Review by nickmcmahon01 — APPROVED** [review]
Approved, but did we not agree to add the permissions check to the method(s) that are missing them at the moment?

File: src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java (hunk 1 above)
Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java]: Changed...

File: src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java (hunk 3 above)
Comment (LiamMacP) [reply src/main/java/uk/gov/digital/ho/hocs/casework/security/UserPermissionsService.java:21]: Nee noor nee noor 👮🏽

**Pull request merged**